base-notebook/start.sh

```diff
@@ -11,140 +12,185 @@ else
-                rm /tmp/passwd
-            else
-                echo 'Container must be run with group "root" to update passwd file'
-            fi
-        fi
+    # Warn about misconfiguration of: desired username, user id, or group id
+    if [[ -n "${NB_USER}" && "${NB_USER}" != "$(id -un)" ]]; then
+        echo "WARNING: container must be started as root to change the desired user's name with NB_USER!"
+    fi
+    if [[ -n "${NB_UID}" && "${NB_UID}" != "$(id -u)" ]]; then
+        echo "WARNING: container must be started as root to change the desired user's id with NB_UID!"
+    fi
+    if [[ -n "${NB_GID}" && "${NB_GID}" != "$(id -g)" ]]; then
+        echo "WARNING: container must be started as root to change the desired user's group id with NB_GID!"
+    fi
 
-        # Warn if the user isn't going to be able to write files to ${HOME}.
-        if [[ ! -w /home/jovyan ]]; then
-            echo 'Container must be run with group "users" to update files'
-        fi
-    else
-        # Warn if looks like user want to override uid/gid but hasn't
-        # run the container as root.
-        if [[ -n "${NB_UID}" && "${NB_UID}" != "$(id -u)" ]]; then
-            echo "Container must be run as root to set NB_UID to ${NB_UID}"
-        fi
-        if [[ -n "${NB_GID}" && "${NB_GID}" != "$(id -g)" ]]; then
-            echo "Container must be run as root to set NB_GID to ${NB_GID}"
+    # Warn about misconfiguration of: granting sudo rights
+    if [[ "${GRANT_SUDO}" == "1" || "${GRANT_SUDO}" == 'yes' ]]; then
+        echo "WARNING: container must be started as root to grant sudo permissions!"
+    fi
+
+    # Attempt to ensure the user uid we currently run as has a named entry in
+    # the /etc/passwd file, as it avoids software crashing on hard assumptions
+    # on such entry. Writing to the /etc/passwd was allowed for the root group
+    # from the Dockerfile during build.
+    #
+    # ref: https://github.com/jupyter/docker-stacks/issues/552
+    if ! whoami &> /dev/null; then
+        if [[ -w /etc/passwd ]]; then
+            sed --in-place "s/^jovyan:/nayvoj:/" /etc/passwd
+            echo "Renamed old jovyan user to nayvoy (1000:100)"
+
+            echo "jovyan:x:$(id -u):$(id -g):,,,:/home/jovyan:/bin/bash" >> /etc/passwd
+            echo "Added new jovyan user ($(id -u):$(id -g))"
+        else
+            echo "WARNING: container must be started with group 'root' (0) to add a user entry in /etc/passwd!"
         fi
     fi
 
-    # Warn if looks like user want to run in sudo mode but hasn't run
-    # the container as root.
-    if [[ "${GRANT_SUDO}" == "1" || "${GRANT_SUDO}" == 'yes' ]]; then
-        echo 'Container must be run as root to grant sudo permissions'
+    # Warn if the user isn't able to write files to ${HOME}
+    if [[ ! -w /home/jovyan ]]; then
+        echo "WARNING: container must be started with group 'users' (100) to be allowed to write to /home/jovyan!"
     fi
```

Comment:
Moved above to [here](https://github.com/jupyter/docker-stacks/pull/1512/files#r744325696)